File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/AttachmentStore.kt
package dev.meeterm.terminal

import android.content.Context
import android.net.Uri
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * App-owned attachment files under `cacheDir/attachments/`.
 *
 * Sources are stream-copied through a bounded buffer; the picker never hands
 * the pipeline an unbounded byte array. Names are app-generated so provider
 * filenames can never traverse or collide inside the directory.
 */
internal class AttachmentStore(private val context: Context) {
  private val random = SecureRandom()

  val directory: File
    get() = File(context.cacheDir, "attachments")

  private fun newFileName(extension: String): String {
    val bytes = ByteArray(12).also { random.nextBytes(it) }
    val token = bytes.joinToString("") { "%02x".format(it) }
    return "att_$token.$extension"
  }

  fun newStagingFileName(): String = newFileName("bin")

  fun newPreparedFileName(format: AttachmentImageFormat): String = newFileName(format.extension)

  fun resolveAppFile(name: String): File? {
    if (!AttachmentFileNames.isValid(name)) return null
    val file = File(directory, name)
    return if (file.parentFile?.canonicalPath == directory.canonicalPath) file else null
  }

  fun stagingFile(name: String): File? = resolveAppFile(name)?.takeIf {
    AttachmentFileNames.isStagingName(name)
  }

  fun preparedFile(name: String): File? = resolveAppFile(name)?.takeIf {
    AttachmentFileNames.isPreparedName(name)
  }

  sealed class StageCopy {
    data class Ok(val fileName: String, val byteCount: Long) : StageCopy()
    data class Rejected(val errorCode: String) : StageCopy()
  }

  /** Bounded Uri → staging-file copy; refuses early on size overruns. */
  fun stageFromUri(uri: Uri, fileName: String): StageCopy {
    val target = stagingFile(fileName) ?: return StageCopy.Rejected(
      AttachmentLimits.ERROR_ARGUMENT,
    )
    directory.mkdirs()
    return try {
      context.contentResolver.openInputStream(uri).use { input ->
        if (input == null) return StageCopy.Rejected(AttachmentLimits.ERROR_IO)
        var total = 0L
        val buffer = ByteArray(STREAM_BUFFER_BYTES)
        FileOutputStream(target).use { output ->
          while (true) {
            val read = input.read(buffer)
            if (read < 0) break
            total += read
            if (total > AttachmentLimits.MAX_SOURCE_BYTES) {
              output.close()
              target.delete()
              return StageCopy.Rejected(AttachmentLimits.ERROR_INPUT_TOO_LARGE)
            }
            output.write(buffer, 0, read)
          }
        }
        if (total <= 0L) {
          target.delete()
          StageCopy.Rejected(AttachmentLimits.ERROR_IO)
        } else {
          StageCopy.Ok(fileName, total)
        }
      }
    } catch (e: SecurityException) {
      target.delete()
      StageCopy.Rejected(AttachmentLimits.ERROR_IO)
    } catch (e: IOException) {
      target.delete()
      StageCopy.Rejected(AttachmentLimits.ERROR_IO)
    }
  }

  fun readHeader(fileName: String): Pair<ByteArray, Long>? {
    val file = stagingFile(fileName) ?: return null
    if (!file.isFile) return null
    val length = file.length()
    return try {
      FileInputStream(file).use { input ->
        val buffer = ByteArray(AttachmentLimits.HEADER_SNIFF_BYTES)
        var offset = 0
        while (offset < buffer.size) {
          val read = input.read(buffer, offset, buffer.size - offset)
          if (read < 0) break
          offset += read
        }
        buffer.copyOf(offset) to length
      }
    } catch (e: IOException) {
      null
    }
  }

  fun stagingPath(fileName: String): File? = stagingFile(fileName)?.takeIf { it.isFile }

  fun preparedPath(fileName: String): File? = preparedFile(fileName)?.takeIf { it.isFile }

  fun previewUri(fileName: String): String {
    val file = preparedFile(fileName) ?: return ""
    return Uri.fromFile(file).toString()
  }

  fun sha256(fileName: String, staging: Boolean): String? {
    val file = (if (staging) stagingFile(fileName) else preparedFile(fileName)) ?: return null
    if (!file.isFile) return null
    val digest = MessageDigest.getInstance("SHA-256")
    return try {
      FileInputStream(file).use { input ->
        val buffer = ByteArray(STREAM_BUFFER_BYTES)
        while (true) {
          val read = input.read(buffer)
          if (read < 0) break
          digest.update(buffer, 0, read)
        }
        digest.digest().joinToString("") { "%02x".format(it) }
      }
    } catch (e: IOException) {
      null
    }
  }

  fun delete(vararg names: String?) {
    for (name in names) {
      val file = name?.let { resolveAppFile(it) } ?: continue
      if (file.isFile) file.delete()
    }
  }

  /**
   * Startup reclaim of abandoned files from a previous process. Anything not
   * referenced by the live session goes away; an active session's files stay.
   */
  fun reclaimStale(keptNames: Set<String>) {
    val dir = directory
    val files = dir.listFiles() ?: return
    for (file in files) {
      if (file.name in keptNames) continue
      // Unrecognized files are reclaimed too: only the app writes here.
      file.delete()
    }
  }

  companion object {
    private const val STREAM_BUFFER_BYTES = 64 * 1024
  }
}

File: modules/meeterm-terminal/ios/AttachmentStore.swift
import Foundation

/**
 * App-owned attachment files under `Caches/attachments/`.
 *
 * Sources stream through a bounded buffer; the picker never hands the
 * pipeline an unbounded Data. Names are app-generated so provider filenames
 * can never traverse or collide inside the directory.
 */
final class AttachmentStore {
  enum StageCopy {
    case ok(fileName: String, byteCount: Int64)
    case rejected(String)
  }

  private let fileManager = FileManager.default

  let directoryURL: URL

  init() throws {
    let base = try fileManager.url(
      for: .cachesDirectory,
      in: .userDomainMask,
      appropriateFor: nil,
      create: true
    )
    directoryURL = base.appendingPathComponent("attachments", isDirectory: true)
    try fileManager.createDirectory(
      at: directoryURL,
      withIntermediateDirectories: true
    )
  }

  init(directoryURL: URL) throws {
    self.directoryURL = directoryURL
    try fileManager.createDirectory(
      at: directoryURL,
      withIntermediateDirectories: true
    )
  }

  private func newFileName(_ fileExtension: String) -> String {
    let token = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
    return "att_\(String(token.prefix(24))).\(fileExtension)"
  }

  func newStagingFileName() -> String { newFileName("bin") }

  func newPreparedFileName(_ format: AttachmentImageFormat) -> String {
    newFileName(format.fileExtension)
  }

  func resolveAppFile(_ name: String) -> URL? {
    guard AttachmentFileNames.isValid(name) else { return nil }
    let url = directoryURL.appendingPathComponent(name, isDirectory: false)
    guard url.deletingLastPathComponent().standardizedFileURL.path ==
          directoryURL.standardizedFileURL.path else { return nil }
    return url
  }

  func stagingFile(_ name: String) -> URL? {
    guard AttachmentFileNames.isStagingName(name) else { return nil }
    return resolveAppFile(name)
  }

  func preparedFile(_ name: String) -> URL? {
    guard AttachmentFileNames.isPreparedName(name) else { return nil }
    return resolveAppFile(name)
  }

  /// Bounded URL → staging-file copy; refuses early on size overruns.
  func stage(from sourceURL: URL, fileName: String, securityScoped: Bool) -> StageCopy {
    guard let target = stagingFile(fileName) else {
      return .rejected(AttachmentLimits.errorArgument)
    }
    if securityScoped, !sourceURL.startAccessingSecurityScopedResource() {
      return .rejected(AttachmentLimits.errorIO)
    }
    defer {
      if securityScoped { sourceURL.stopAccessingSecurityScopedResource() }
    }
    guard let input = InputStream(url: sourceURL) else {
      return .rejected(AttachmentLimits.errorIO)
    }
    input.open()
    defer { input.close() }
    guard let output = OutputStream(url: target, append: false) else {
      return .rejected(AttachmentLimits.errorIO)
    }
    output.open()
    defer { output.close() }
    var total: Int64 = 0
    var buffer = [UInt8](repeating: 0, count: 64 * 1024)
    while true {
      let read = input.read(&buffer, maxLength: buffer.count)
      if read < 0 { try? fileManager.removeItem(at: target); return .rejected(AttachmentLimits.errorIO) }
      if read == 0 { break }
      total += Int64(read)
      if total > AttachmentLimits.maxSourceBytes {
        try? fileManager.removeItem(at: target)
        return .rejected(AttachmentLimits.errorInputTooLarge)
      }
      var written = 0
      while written < read {
        let amount = buffer.withUnsafeBytes { raw -> Int in
          guard let base = raw.baseAddress?.advanced(by: written) else { return -1 }
          return output.write(base.assumingMemoryBound(to: UInt8.self), maxLength: read - written)
        }
        if amount <= 0 {
          try? fileManager.removeItem(at: target)
          return .rejected(AttachmentLimits.errorIO)
        }
        written += amount
      }
    }
    guard total > 0 else {
      try? fileManager.removeItem(at: target)
      return .rejected(AttachmentLimits.errorIO)
    }
    return .ok(fileName: fileName, byteCount: total)
  }

  /// Move a picker-provided temp file into staging with a size check first.
  func stageFile(_ sourceURL: URL, fileName: String) -> StageCopy {
    guard let target = stagingFile(fileName) else {
      return .rejected(AttachmentLimits.errorArgument)
    }
    do {
      let attributes = try fileManager.attributesOfItem(atPath: sourceURL.path)
      let size = (attributes[.size] as? NSNumber)?.int64Value ?? 0
      guard size > 0 else { return .rejected(AttachmentLimits.errorIO) }
      guard size <= AttachmentLimits.maxSourceBytes else {
        return .rejected(AttachmentLimits.errorInputTooLarge)
      }
      try? fileManager.removeItem(at: target)
      try fileManager.moveItem(at: sourceURL, to: target)
      return .ok(fileName: fileName, byteCount: size)
    } catch {
      return .rejected(AttachmentLimits.errorIO)
    }
  }

  func readHeader(_ fileName: String) -> (header: Data, byteCount: Int64)? {
    guard let url = stagingFile(fileName),
          fileManager.fileExists(atPath: url.path),
          let size = (try? fileManager.attributesOfItem(atPath: url.path))?[.size] as? NSNumber,
          let handle = try? FileHandle(forReadingFrom: url) else { return nil }
    defer { try? handle.close() }
    let header = (try? handle.read(upToCount: AttachmentLimits.headerSniffBytes)) ?? Data()
    return (header, size.int64Value)
  }

  func stagingURL(_ fileName: String) -> URL? {
    guard let url = stagingFile(fileName), fileManager.fileExists(atPath: url.path) else { return nil }
    return url
  }

  func preparedURL(_ fileName: String) -> URL? {
    guard let url = preparedFile(fileName), fileManager.fileExists(atPath: url.path) else { return nil }
    return url
  }

  func previewUri(_ fileName: String) -> String {
    guard let url = preparedFile(fileName) else { return "" }
    return url.absoluteString
  }

  func delete(_ names: String?...) {
    for name in names {
      guard let name = name, let url = resolveAppFile(name) else { continue }
      try? fileManager.removeItem(at: url)
    }
  }

  /// Startup reclaim of abandoned files from a previous process.
  func reclaimStale(keeping keptNames: Set<String>) {
    guard let contents = try? fileManager.contentsOfDirectory(
      at: directoryURL,
      includingPropertiesForKeys: nil
    ) else { return }
    for url in contents where !keptNames.contains(url.lastPathComponent) {
      try? fileManager.removeItem(at: url)
    }
  }
}
